File: ios/Sources/ForgeIOSUI/StableTabsRenderer.swift
import SwiftUI
import ForgeIOSRuntime

struct StableTabsRenderer: View {
    let runtime: ForgeRuntime?
    let window: WindowContext?
    let container: ContainerDef
    let form: [String: JSONValue]
    let collection: [[String: JSONValue]]
    let metrics: [String: JSONValue]
    let windowForm: [String: JSONValue]
    let selection: SelectionState
    @State private var selectedID: String?
    @State private var visited = Set<String>()
    @State private var authorization: [String: JSONValue] = [:]
    @State private var grants: [String: [[String: JSONValue]]] = [:]
    private var stateKey: String { "__forgeStableTab:\(container.id ?? container.containers.first?.id ?? "root")" }
    private func allows(_ condition: DashboardConditionDef?) -> Bool {
        DashboardRuntime.evaluateDashboardCondition(condition, metrics: metrics.mapValues(unwrap), form: form.mapValues(unwrap), windowForm: windowForm.mapValues(unwrap), collection: collection.map { $0.mapValues(unwrap) })
    }
    private var visible: [ContainerDef] {
        container.containers.filter { child in
            guard allows(child.visibleWhen) else { return false }
            guard let permission = child.permissionBoundary else { return true }
            guard allows(permission.visibleWhen) else { return false }
            let rows = permission.mode == "selection" ? (selection.selection.isEmpty ? selection.selected.map { [$0] } ?? [] : selection.selection) : permission.mode == "row" ? collection : []
            return WorkflowPrimitiveRuntime.permissionAllows(spec: permission, authorization: authorization, rows: rows, grants: grants[permission.dataSourceRef ?? ""] ?? [])
        }
    }
    var body: some View {
        let ids = visible.compactMap(\.id)
        let selected = StableTabsState.selected(ids: ids, requested: selectedID ?? windowForm[stateKey]?.stringValue, fallback: container.stableTabs?.defaultSelectedTabId)
        let mounted = StableTabsState.mounted(ids: ids, selected: selected, visited: visited, keepVisited: container.stableTabs?.keepVisitedTabPanelsMounted == true, activeOnly: container.stableTabs?.renderActiveTabPanelOnly != false)
        VStack(alignment: .leading, spacing: 8) {
            if let selected {
                Picker(container.title ?? "Tabs", selection: Binding(get: { selected }, set: select)) {
                    ForEach(visible) { child in Text(child.title ?? child.id ?? "Tab").tag(child.id ?? "") }
                }.accessibilityLabel(container.title ?? "Tabs")
                ForEach(visible.filter { mounted.contains($0.id ?? "") }) { child in
                    ContainerRenderer(runtime: runtime, window: window, container: child, inheritedDataSourceRef: container.dataSourceRef, suppressTitle: true)
                        .frame(height: child.id == selected ? nil : 0)
                        .opacity(child.id == selected ? 1 : 0)
                        .clipped().disabled(child.id != selected).accessibilityHidden(child.id != selected)
                }
            }
        }
        .onAppear { if let selected { visited.insert(selected) } }
        .onChange(of: ids) { if let next = StableTabsState.selected(ids: ids, requested: selectedID ?? windowForm[stateKey]?.stringValue, fallback: container.stableTabs?.defaultSelectedTabId) { select(next) } }
        .onChange(of: windowForm[stateKey]) { selectedID = windowForm[stateKey]?.stringValue; if let selectedID { visited.insert(selectedID) } }
        .task(id: container.containers.compactMap { $0.permissionBoundary?.dataSourceRef }.joined(separator: "|")) {
            guard let runtime, let window else { return }
            authorization = await runtime.windowMetadata(id: window.windowID)?.authorizationSnapshot ?? [:]
            let refs = Set(container.containers.compactMap { $0.permissionBoundary?.dataSourceRef })
            for ref in refs { grants[ref] = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: ref) }
            await withTaskGroup(of: Void.self) { group in
                group.addTask {
                    for await metadata in await runtime.windowMetadataUpdates(id: window.windowID) { await MainActor.run { authorization = metadata?.authorizationSnapshot ?? [:] } }
                }
                for ref in refs { group.addTask { for await value in await runtime.dataSourceCollectionUpdates(windowID: window.windowID, dataSourceRef: ref) { await MainActor.run { grants[ref] = value } } } }
            }
        }
    }
    private func select(_ id: String) {
        let revisit = visited.contains(id)
        selectedID = id; visited.insert(id)
        guard let runtime, let window else { return }
        Task {
            await runtime.setWindowFormValue(windowID: window.windowID, values: [stateKey: .string(id)], bumpPrefillRevision: false)
            if revisit && container.stableTabs?.dataSourceFetchMode != "once", let ref = visible.first(where: { $0.id == id })?.dataSourceRef ?? container.dataSourceRef {
                await runtime.refreshDataSourceCollection(windowID: window.windowID, dataSourceRef: ref)
            }
        }
    }
    private func unwrap(_ value: JSONValue) -> Any {
        switch value { case .string(let v): return v; case .number(let v): return v; case .bool(let v): return v; case .object(let v): return v.mapValues(unwrap); case .array(let v): return v.map(unwrap); case .null: return NSNull() }
    }
}

File: android/sdk/src/main/java/com/viant/forgeandroid/ui/StableTabsRenderer.kt
package com.viant.forgeandroid.ui

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clipToBounds
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.unit.dp
import com.viant.forgeandroid.runtime.*
import kotlinx.serialization.json.JsonPrimitive

@Composable
internal fun StableTabsRenderer(runtime: ForgeRuntime, window: WindowContext, container: ContainerDef, inheritedDataSourceRef: String?) {
    val focusManager = androidx.compose.ui.platform.LocalFocusManager.current
    val windowForm by window.windowFormSignal().flow.collectAsState(initial = window.peekWindowForm())
    val source = (container.dataSourceRef ?: inheritedDataSourceRef)?.let(window::contextOrNull)
    val form by source?.form?.flow?.collectAsState(initial = source.form.peek()) ?: remember { mutableStateOf(emptyMap()) }
    val collection by source?.collection?.flow?.collectAsState(initial = source.collection.peek()) ?: remember { mutableStateOf(emptyList()) }
    val metrics by source?.metrics?.flow?.collectAsState(initial = source.metrics.peek()) ?: remember { mutableStateOf(emptyMap()) }
    val selection by source?.selection?.flow?.collectAsState(initial = source.selection.peek()) ?: remember { mutableStateOf(SelectionState()) }
    val metadata by window.metadata.flow.collectAsState(initial = window.metadata.peek())
    val authorization = metadata?.authorizationSnapshot?.mapValues { JsonUtil.elementToAny(it.value) }.orEmpty()
    fun allows(condition: DashboardConditionDef?) = evaluateDashboardCondition(condition, metrics = metrics, form = form, windowForm = windowForm, collection = collection, authorization = authorization)
    val visible = mutableListOf<ContainerDef>()
    for (child in container.containers) {
        val permission = child.permissionBoundary
        val permissionSource = permission?.dataSourceRef?.let(window::contextOrNull)
        val grants by permissionSource?.collection?.flow?.collectAsState(initial = permissionSource.collection.peek()) ?: remember { mutableStateOf(emptyList()) }
        val rows = when (permission?.mode) { "selection" -> selection.selection.ifEmpty { selection.selected?.let(::listOf).orEmpty() }; "row" -> collection; else -> emptyList() }
        if (allows(child.visibleWhen) && (permission == null || (allows(permission.visibleWhen) && WorkflowPrimitiveRuntime.permissionAllows(permission, authorization, rows, grants)))) visible += child
    }
    val ids = visible.mapNotNull { it.id }
    val stateKey = "__forgeStableTab:${container.id ?: container.containers.firstOrNull()?.id ?: "root"}"
    val selected = StableTabsState.selected(ids, windowForm[stateKey]?.toString(), container.stableTabs?.defaultSelectedTabId)
    var visited by remember(window.windowId, container.id) { mutableStateOf(emptySet<String>()) }
    val mounted = StableTabsState.mounted(ids, selected, visited, container.stableTabs?.keepVisitedTabPanelsMounted == true, container.stableTabs?.renderActiveTabPanelOnly != false)
    LaunchedEffect(selected) { focusManager.clearFocus(); if (selected != null) visited = visited + selected }
    fun select(id: String) {
        if (id in visited && container.stableTabs?.dataSourceFetchMode != "once") (visible.firstOrNull { it.id == id }?.dataSourceRef ?: container.dataSourceRef)?.let { runtime.refreshDataSourceCollection(window.windowId, it) }
        windowForm[stateKey]?.toString().takeIf { it == id } ?: runtime.setWindowFormValues(window.windowId, mapOf(stateKey to id), bumpPrefillRevision = false)
        visited = visited + id
    }
    if (selected == null) return
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        var expanded by remember { mutableStateOf(false) }
        Box {
            OutlinedButton(onClick = { expanded = true }) { Text(visible.first { it.id == selected }.title ?: selected) }
            DropdownMenu(expanded, onDismissRequest = { expanded = false }) { visible.forEach { child -> DropdownMenuItem(text = { Text(child.title ?: child.id.orEmpty()) }, onClick = { child.id?.let(::select); expanded = false }) } }
        }
        visible.filter { it.id in mounted }.forEach { child -> key(child.id) {
            val active = child.id == selected
            Box(if (active) Modifier else Modifier.height(0.dp).alpha(0f).clipToBounds().clearAndSetSemantics {}) {
                ContainerRenderer(runtime, window, child, inheritedDataSourceRef = container.dataSourceRef ?: inheritedDataSourceRef, suppressTitle = true)
            }
        } }
    }
}
